dnprogs/libdnet/dnet_addr.3
.TH DNET_ADDR 3 "July 28, 1998" "DECnet database functions"
.SH NAME
dnet_addr \- DECnet nodename to address translation
.SH SYNOPSIS
.B #include <netdnet/dn.h>
.br
.B #include <netdnet/dnetdb.h>
.br
.sp
.B struct dn_naddr *dnet_addr (char *nodename)
.sp
.SH DESCRIPTION

.B dnet_addr
search the decnet hosts file for 
.B nodename 
and returns the DECnet address in the 
.B dn_naddr
structure
.br
If no entry is found, returns 
.B NULL


.SH EXAMPLE
.nf

#include <netdnet/dn.h>
#include <netdnet/dnetdb.h>
#include <sys/socket.h>

main(void)
{
    struct dn_naddr		*binaddr;
    struct sockaddr_dn		sockaddr;
    int				sockfd;

    if ( (binaddr=dnet_addr("mv3100")) == NULL)
    {
	   printf("No entry in /etc/decnet.conf for mv3100\\n");
	   exit(0);
     }
     if ((sockfd=socket(AF_DECnet,SOCK_SEQPACKET,DNPROTO_NSP)) == \-1) 
     {
    		perror("socket");
    		exit(\-1);
     }

     sockaddr.sdn_family 	= AF_DECnet;
     sockaddr.sdn_flags		= 0x00;
     sockaddr.sdn_objnum	= 0x19;			/* MIRROR */
     sockaddr.sdn_objnamel	= 0x00;
     memcpy(sockaddr.sdn_add.a_addr, binaddr->a_addr,binaddr->a_len);

     if (connect(sockfd, (struct sockaddr *)&sockaddr, 
		sizeof(sockaddr)) < 0) 
     {
		perror("connect");
		exit(\-1);
     }
     close(sockfd);
}
.fi




.SH SEE ALSO

.BR dnet_htoa (3),
.BR dnet_ntoa (3),
.BR dnet_conn (3),
.BR getnodeadd (3),
.BR getnodebyname (3),
.BR getnodebyaddr (3),
.BR setnodeent (3)


dnprogs/NEWS
dnprogs (2.61) unstable; urgency=low

  * Include libvaxdata/src/test.c in the distribution tarball
  * Fix password length in DAP connections. It was incorrectly set at 12 
    rather than 40
 
 -- Chrissie Caulfield <chrissie@debian.org>  Sun, 31 Mar 2013 11:18:32 +0000

dnprogs (2.60+x32) unreleased; urgency=low

  * Switch from using deprecated sysctl(2) to writing to /proc/sys.  This
    fixes a build failure on x32.

 -- Daniel Schepler <schepler@debian.org>  Tue, 12 Feb 2013 12:21:31 -0800

dnprogs (2.60) unstable; urgency=low

  * Fixed lots of Debian things that caused lintian problems and 
    upgraded policy version
    Closes: #673408
  * Fix some warnings that might be bugs
  * Incorporate NMU into main tree.

 -- Chrissie Caulfield <chrissie@debian.org>  Sat, 26 May 2012 09:54:41 +0100

dnprogs (2.59+nmu2) unstable; urgency=medium

  * NMU
  * dnprogs binaries are Linux-only, so state that. Move dnet-common back
    to binary-all. Closes: #665886

 -- Steve McIntyre <93sam@debian.org>  Mon, 21 May 2012 17:43:14 +0100

dnprogs (2.59+nmu1) unstable; urgency=low

  * NMU
  * Reduce libdnet's recommends on dnet-common to suggests again.
    Closes: #655740. Adapted patch from Jonathan Nieder.

 -- Steve McIntyre <93sam@debian.org>  Fri, 18 May 2012 13:29:53 +0100

dnprogs (2.59) unstable; urgency=low

  * Changed package dnet-common from arch all to any and
    updated dependecies. (Closes: #665886)

 -- Chrissie Caulfield <chrissie@debian.org>  Fri, 27 Apr 2012 09:30:19 +0100

dnprogs (2.58) unstable; urgency=medium

  * Fixed lintian warnings:
    - Fixed changelog entry of last release to have correct line length.
    - Corrected debconf dependency in dnet-common.
    - Corrected versioned dependency in dnprogs on dnet-common.
    - Updated existing overrides to match new syntax.
  * Updated linkage of libdnet_daemon and libdap. (Closes: #646353)
  * Fixed compile warning in dndir when using -Werror=format-security
    (Closes: #646613)
  * Fixed typo in debian/dnet-common.config
  * Do not try to install files into packages on FreeBSD which are
    not build on FreeBSD.
  * Link dapfs with libpthread for new libfuse compatibility
  * Updated nl.po (Closes: #652349)
  * Added pl.po (Closes: #661894)
  * Added it.po (Closes: #663135)
  * Change module-init-tools dependency to kmod (Closes: #662684)

 -- Chrissie Caulfield <chrissie@debian.org>  Sat, 10 Mar 2012 10:32:08 +0000

dnprogs (2.57) unstable; urgency=medium

  [ Chrissie Caulfield ]
  * Completed removal of modutils configuration file code
  * Fix presence of CVS directories in tarball
  * updated 'clean' target

  [ Philipp Schafft ]
  * Use /dev/pts PTYs on 64 bit systems (fix makefile bug)
  * Updated translations Debconf templates:
    - German translation (Closes: #615268)
    - French translation (Closes: #615039)
    - Danish translation (Closes: #633984) (added)
    - Slovak translation (Closes: #639629) (added)
  * Look in yet another place for libcrypt (better fix than from NMUs)
  * Added a comment about the detecting of libcrypt. (See: #629664)
  * "ethernet" was capitalized in extended description (Closes: #636798)
  * Fixed some lintian warnings (mainly new rules targets).
  * dnet-common: Do not configure DECnet unless the user requested it.
    (Removed the bad workaround done by NMU) (See: #637179)
  * dnet-common: Added preinst script to clean up the broken config files
    from the last NMU. (Closes: #640861)
  * Upgraded libdnet's suggest to dnet-common to recommend again
    (Closes: #608807)

 -- Chrissie Caulfield <chrissie@debian.org>  Thu,  8 Sep 2011 16:05:47 +0200

dnprogs (2.56.1+nmu1) unstable; urgency=low

  * Non-maintainer upload.
  * libdnet: Only suggest dnet-common package (Closes: #608807)
  * dnet-common: Do not change MAC addresses unless user chooses to
    configure DECnet (Closes: #637179)
  * dnet-common: Restore ability to configure DECnet name and address
    through debconf
  * dnet-common: Remove obsolete modutils configuration file

 -- Ben Hutchings <ben@decadent.org.uk>  Thu, 25 Aug 2011 14:26:12 +0100

dnprogs (2.56.1) unstable; urgency=low

  * Non-maintainer upload.
  * Add patch from Ubuntu to fix FTBFS with libcrypt moved to the
    multiarch path.

 -- Aurelien Jarno <aurel32@debian.org>  Tue, 09 Aug 2011 11:58:06 +0200


dnprogs (2.56) unstable; urgency=low

  * Search for libgcrypt in 64bit library paths, too. (thanks to Kamil)
  * Package dnet-common now asks if network interfaces should be configrued
    for DECnet (Closes: #608807)
    Updated translations Debconf templates:
    - Russian translation (Closes: #610642)
    - Portuguese translation (Closes: #611034)
    - Swedish translation (Closes: #611296)
    - Czech translation (Closes: #611430)
    - Japanese translation (Closes: #611448)
    - Spanish translation (Closes: #611547)
  * Build against version 25 of FUSE
    Closes: #598831
  * Fixed some lintian warnings (typos and such).
  * Raised priority of package libdnet from extra to optional
    to comply policy manual section 2.5 "Priorities".

 -- Chrissie Caulfield <chrissie@debian.org>  Tue, 22 Feb 2011 10:38:23 +0000

dnprogs (2.55) unstable; urgency=low

  * Build against version 25 of FUSE
    Closes: #598831

 -- Chrissie Caulfield <chrissie@debian.org>  Tue, 5 Oct 2010 09:23:12 +0100

dnprogs (2.54) unstable; urgency=low

  * Don't build libvaxdata test program by default.
    Closes: #598752
 
 -- Chrissie Caulfield <chrissie@debian.org>  Fri, 1 Oct 2010 20:44:09 +0100

dnprogs (2.53) unstable; urgency=low

  * Done some more porting to FreeBSD, corrected Build-Depends for hurd-i386
    Closes: #588185
  * Added new tool node(1) - Node name lookup tool
  * Added support to libdnet_daemon to read /etc/nodes.{allow,deny}
  * Fixed DoS bug in dnet_daemon()
    Closes: #592015 
  * Update libvaxdata to version 1.2, fixes double conversions

 -- Christine Caulfield <Christine.Caulfield@googlemail.com>  Thu, 19 Jul 2010 10:19:32 +0100


dnprogs/libdnet/dnet_addr.c
/******************************************************************************
    (c) 1995-1998 E.M. Serrat          emserrat@geocities.com

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#include <stdio.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <errno.h>
#include <string.h>
#include <stdlib.h>
#include <netdnet/dn.h>
#include <netdnet/dnetdb.h>

static char             nodetag[80],nametag[80],nodeadr[80],nodename[80];
static struct dn_naddr	binadr = {0x0002,{0x00,0x00}};


struct	dn_naddr	*dnet_addr(char *name)
{

	FILE		*dnhosts;
	char		nodeln[80];
        char            **endptr;
	char		*aux;
	long		area,node;


	if ((dnhosts = fopen(SYSCONF_PREFIX "/etc/decnet.conf","r")) == NULL)
	{
		printf("dnet_addr: Can not open " SYSCONF_PREFIX "/etc/decnet.conf\n");
		errno = ENOENT;
		return NULL;
	}
	while (fgets(nodeln,80,dnhosts) != NULL)
	{
		sscanf(nodeln,"%s%s%s%s\n",nodetag,nodeadr,nametag,nodename);
		if (strncmp(nodetag,"#",1) != 0)
		{
		   if (((strcmp(nodetag,"executor") != 0) &&
	    	       (strcmp(nodetag,"node")     != 0)) ||
		       (strcmp(nametag,"name")     != 0))
		   {
		       printf("dnet_addr: Invalid decnet.conf syntax\n");
		       errno = ENOENT;
		       return 0;
		   }
		   if (strcmp(nodename,name) == 0)
		   {
			aux=nodeadr;
			endptr=&aux;
			area=strtol(nodeadr,endptr,0);
			node=strtol(*endptr+1,endptr,0);
			if ((area < 0) || (area > 63) ||
			    (node < 0) || (node > 1023))
			{
				printf("dnet_addr: Invalid address %d.%d\n",
							(int)area,(int)node);
				fclose(dnhosts);
				return 0;
			}
			binadr.a_addr[0] = node & 0xFF;
			binadr.a_addr[1] = (area << 2) | ((node & 0x300) >> 8);
			fclose(dnhosts);
			return &binadr;
		   }
		}
	}
	fclose(dnhosts);
	errno = ENOENT;
	return 0;
}
